### PR TITLE
Add surface particles intialization to ParticleInit.

### PR DESCRIPTION
**Pull request by abisner**
This PR adds a method to generate particles on a faceted surface (e.g. one produced by marching cubes). Currently only random initialization.

### CODE REVIEW

**Comment by abisner** [comment]
This should be ready to merge.

**Review by streeve — CHANGES_REQUESTED** [review]
A few minor changes - I'd also like to walk through the algorithm with you quickly

File: src/Picasso_ParticleInit.hpp [hunk 1]
```diff
@@ -368,6 +369,142 @@ void initializeParticles( InitUniform, const ExecutionSpace& exec_space,
...
+    int num_facets = surface.facets.extent( 0 );
+    int num_particles = particles_per_facet * num_facets;
+    particles.resize( num_particles );
+
+    // Creation status.
+    auto particle_created = Kokkos::View<bool*, memory_space>(
+        Kokkos::ViewAllocateWithoutInitializing( "particle_created" ),
+        num_particles );
+
+    // Create a random number generator.
```
Comment (streeve) [inline src/Picasso_ParticleInit.hpp:441]: From discussion: add issue for shared random generation utiltiy

File: src/Picasso_ParticleInit.hpp [hunk 2]
```diff
@@ -368,6 +369,142 @@ void initializeParticles( InitUniform, const ExecutionSpace& exec_space,
...
+    uint64_t seed =
+        global_grid.blockId() + ( 19383747 % ( global_grid.blockId() + 1 ) );
+    using rnd_type = Kokkos::Random_XorShift64_Pool<ExecutionSpace>;
+    rnd_type pool;
+    pool.init( seed, num_facets );
+
+    // // Initialize particles.
+    Kokkos::parallel_for(
+        "Picasso::ParticleInit::Uniform",
+        Kokkos::RangePolicy<ExecutionSpace>( 0, num_facets ),
```
Comment (streeve) [inline src/Picasso_ParticleInit.hpp:451]: From discussion: fixme comment for iterating over `num_particles` instead

File: src/Picasso_ParticleInit.hpp [hunk 3]
```diff
@@ -368,6 +369,142 @@ void initializeParticles( InitUniform, const ExecutionSpace& exec_space,
...
+    auto& particles = particle_list.aosoa();
+
+    // Allocate enough space for the case the particles consume the entire
+    // local grid.
+    int num_facets = surface.facets.extent( 0 );
+    int num_particles = particles_per_facet * num_facets;
+    particles.resize( num_particles );
+
+    // Creation status.
+    auto particle_created = Kokkos::View<bool*, memory_space>(
```
[the remark this answers is not in the document]
Comment (abisner) [reply src/Picasso_ParticleInit.hpp]: Removed.

File: src/Picasso_ParticleInit.hpp [hunk 4]
```diff
@@ -368,6 +369,142 @@ void initializeParticles( InitUniform, const ExecutionSpace& exec_space,
...
+    // Create a random number generator.
+    uint64_t seed =
+        global_grid.blockId() + ( 19383747 % ( global_grid.blockId() + 1 ) );
+    using rnd_type = Kokkos::Random_XorShift64_Pool<ExecutionSpace>;
+    rnd_type pool;
+    pool.init( seed, num_facets );
+
+    // // Initialize particles.
+    Kokkos::parallel_for(
+        "Picasso::ParticleInit::Uniform",
```
[the remark this answers is not in the document]
Comment (abisner) [reply src/Picasso_ParticleInit.hpp:447]: Changed label.

File: src/Picasso_ParticleInit.hpp [hunk 5]
```diff
@@ -368,6 +369,142 @@ void initializeParticles( InitUniform, const ExecutionSpace& exec_space,
...
+
+            // Particle.
+            particle_type particle;
+
+            Vec3<double> a = { surface.facets( f, 0, 0 ),
+                               surface.facets( f, 0, 1 ),
+                               surface.facets( f, 0, 2 ) };
+
+            Vec3<double> ab = {
+                surface.facets( f, 1, 0 ) - surface.facets( f, 0, 0 ),
```
[the remark this answers is not in the document]
Comment (abisner) [reply src/Picasso_ParticleInit.hpp]: Done.

File: src/Picasso_ParticleInit.hpp (hunk 2 above)
Comment (abisner) [reply src/Picasso_ParticleInit.hpp:448]: Done.

**Comment by streeve** [comment]
Still some minor unresolved comments, but otherwise ready to approve

File: src/Picasso_ParticleInit.hpp [hunk 6]
```diff
@@ -368,6 +369,142 @@ void initializeParticles( InitUniform, const ExecutionSpace& exec_space,
...
+    // Create a local mesh.
+    auto local_mesh = Cajita::createLocalMesh<ExecutionSpace>( local_grid );
+
+    // Get the global grid.
+    const auto& global_grid = local_grid.globalGrid();
+
+    // Get the particles.
+    auto& particles = particle_list.aosoa();
+
+    // Allocate enough space for the case the particles consume the entire
```
[the remark this answers is not in the document]
Comment (abisner) [reply src/Picasso_ParticleInit.hpp]: Fixed comment.

File: src/Picasso_ParticleInit.hpp [hunk 7]
```diff
@@ -368,6 +369,142 @@ void initializeParticles( InitUniform, const ExecutionSpace& exec_space,
...
+            double facet_area = 0.5 * sqrt( c2 );
+
+            // Particle surface area.
+            double pa = facet_area / particles_per_facet;
+
+            for ( int ip = 0; ip < particles_per_facet; ++ip )
+            {
+                // Local particle id.
+                int pid = f * particles_per_facet + ip;
+                // int pid = ip * particles_per_facet + f;
```
[the remark this answers is not in the document]
Comment (streeve) [reply src/Picasso_ParticleInit.hpp:492]: Shouldn't we remove this commented line?

**Pull request merged**